distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+#!/usr/bin/python
+
+# ----------------------------------------------------------------------------
+# Copyright (c) 2015--, The WGS-HGT Development Team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+""" Implement Distance Method for HGT detection based on algorithm described
+    in:
+        Wei. X et al., "A Distance-Based Method for Detecting HGT in Whole
+        Genomes", International Symposium on Bioinformatics Research and
+        Applications (ISBRA), 2008, pages 26-37
+
+    The workflow follows the algorithm:
+    1. For each gene in target genome,
+        i.    BLAST sequence against all other genes in the reference genomes;
+        ii.   Go to step 3 if gene has more than threshold number of homologs
+              (min-num-homologs), otherwise go to next gene in target genome;
+        iii.  Compute multiple sequence alignment on homolog genes using
+              CLUSTAL;
+        iv.   Compute pairwise distance matrix using PHYLIP's protdist
+              function and Z-score normalize the set of pairwise distances
+              for each gene family and species;
+        v.    Add distance matrix for all pairwise distances into global
+              distance matrix storing results for all genes
+
+    2. Cluster gene families by species,
+        vi.   Compute all species sets (sets of genes whose orthologs are
+              detectable in exactly the same subset of the considered
+              species);
+        vii.  Cluster genes to each core species set cluster using the Hamming
+              distance clustering algorithm;
+        viii. Run outlier detection algorithm on each cluster (paragraph 2
+              of section 'Detecting Outlier Genes' in original paper)
+"""
+
+import sys
+import click
+import numpy
+import operator
+import threading
+import subprocess
+import traceback
+import shlex
+from os import listdir
+from os.path import join, splitext, basename
+from itertools import imap
+
+from skbio.parse.sequences import parse_fasta
+
+
+class Command(object):
+    """
+    Enables to run subprocess commands in a different thread
+    with TIMEOUT option.
+
+    Based on jcollado's solution:
+    http://stackoverflow.com/questions/1191374/subprocess-with-timeout/4825933#4825933
+    https://gist.github.com/kirpit/1306188
+    """
+    command = None
```

Comment:
This is generated at init so you shouldn't need it here. Same with the other variables.